File: utils/download.py
import requests
from utils.getPlaylist import getVideoPlaylist, getStationNo
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry
import time
import os
from urllib.parse import urljoin
from tools.formatBytes import format_bytes
from tools.formatDuration import format_duration
from utils.verify import verify

def download(url, username):
  # attempt to alleviate dropped connections [NEEDS TESTING]
  session = requests.Session()
  retry = Retry(connect=5, backoff_factor=0.5)
  adapter = HTTPAdapter(max_retries=retry)
  session.mount('http://', adapter)
  session.mount('https://', adapter)
  cookie = open('cookies', 'r').read().strip()
  headers = {
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:109.0) Gecko/20100101 Firefox/118.0",
    "Accept": "*/*",
    "Accept-Language": "en-US,en;q=0.5",
    "Accept-Encoding": "gzip, deflate, br",
    "Content-Type": "application/x-www-form-urlencoded",
    "Origin": "https://play.afreecatv.com",
    "DNT": "1",
    "Alt-Used": "live.afreecatv.com",
    "Connection": "keep-alive",
    "Sec-Fetch-Dest": "empty",
    "Sec-Fetch-Mode": "cors",
    "Sec-Fetch-Site": "same-site",
    "Sec-GPC": "1",
    "Cookie": cookie,
    "TE": "trailers",
  }

  base_url = url.rsplit('/', 1)[0] + '/'
  now = time.strftime("%Y-%m-%d_%H:%M", time.localtime())
  output_filename = username + '-' + getStationNo(username,'') + '-' + now + '-afreeca.ts'

  output_path = 'downloads/' + username + '/' + output_filename

  if os.path.exists('downloads/' + username) is False:
      os.makedirs('downloads/' + username)

  segment_urls = set()

  file_size = 0
  start_time = time.time()

  while True:
    response = session.get(url, headers=headers)
    playlist_content = response.text

    if '<p>Sorry, the page you are looking for is currently unavailable.<br/>' in playlist_content:
        print("\nStream paused")
        time.sleep(60)
        continue

    # works for the most part, but needs testing
    if '<title>502 Server Error</title>' in playlist_content:
        print(f"\nStream finished. Exiting the program.")
        if verify(username):
            download(getVideoPlaylist(username, ''), username)

    new_segment_lines = [
        line.strip() for line in playlist_content.splitlines() if line.endswith('.TS')
    ]

    with open(output_path, 'ab') as output_file:
        
        for new_segment_line in new_segment_lines:
            segment_url = urljoin(base_url, new_segment_line)
            if segment_url not in segment_urls:
                segment_urls.add(segment_url)
                response = session.get(segment_url)
                file_size += len(response.content)
                elapsed_time = time.time() - start_time
                output_file.write(response.content)
                print("\r" + f"Downloading to {output_filename} || {format_duration(elapsed_time)} @ {format_bytes(file_size)}             \x1b[?25l", end="", flush=True)

    time.sleep(3)

File: plugins/panda/download.go
package panda

import (
	"bufio"
	"fmt"
	"horsaen/afreeca-downloader/tools"
	"io"
	"net/http"
	"os"
	"strings"
	"time"
)

func ConcurrentDownload(user *[]string, playlist string) {
	userId := (*user)[0]
	tools.Exists("downloads/Panda/" + userId)

	client := &http.Client{}

	cookies := tools.LoadCookies("panda")

	req, _ := http.NewRequest("GET", playlist, nil)

	req.Header.Add("Cookie", "sessKey="+cookies[0])
	req.Header.Add("User-Agent", "Mozilla/5.0 (X11; Linux x86_64; rv:124.0) Gecko/20100101 Firefox/124.0")
	req.Header.Add("Accept", "application/json, text/plain, */*")
	req.Header.Add("Accept-Language", "en-US,en;q=0.5")
	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Add("Origin", "https://www.pandalive.co.kr")
	req.Header.Add("DNT", "1")
	req.Header.Add("Connection", "keep-alive")
	req.Header.Add("Referer", "https://www.pandalive.co.kr/")
	req.Header.Add("Sec-Fetch-Dest", "empty")
	req.Header.Add("Sec-Fetch-Mode", "cors")
	req.Header.Add("Sec-Fetch-Site", "same-site")
	req.Header.Add("Sec-GPC", "1")
	req.Header.Add("TE", "trailers")

	filename := userId + "-" + time.Now().Format("200601021504") + "-panda.ts"

	out, _ := os.Create("downloads/Panda/" + userId + "/" + filename)

	playlistUrls := make(map[string]bool)

	var bytes int64 = 0
	var start_time = time.Now()

	for {
		resp, err := client.Do(req)

		if err != nil {
			(*user)[2] = "ERROR"
			(*user)[3] = "RETRYING"
			(*user)[4] = err.Error()
			ConcurrentDownload(user, playlist)
		}

		bodyBytes, _ := io.ReadAll(resp.Body)

		bodyText := string(bodyBytes)

		if !strings.Contains(bodyText, ".ts") {
			(*user)[2] = "Offline"
			(*user)[3] = "Offline"
			(*user)[4] = "Offline"
			Concurrent(user)
		}

		scanner := bufio.NewScanner(strings.NewReader(bodyText))

		for scanner.Scan() {
			line := scanner.Text()
			if strings.HasPrefix(line, "https://") {
				if !playlistUrls[line] {

					req, _ := http.NewRequest("GET", line, nil)

					req.Header.Add("Cookie", "sessKey="+cookies[0])
					req.Header.Add("User-Agent", "Mozilla/5.0 (X11; Linux x86_64; rv:124.0) Gecko/20100101 Firefox/124.0")
					req.Header.Add("Accept", "application/json, text/plain, */*")
					req.Header.Add("Accept-Language", "en-US,en;q=0.5")
					req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
					req.Header.Add("Origin", "https://www.pandalive.co.kr")
					req.Header.Add("DNT", "1")
					req.Header.Add("Connection", "keep-alive")
					req.Header.Add("Referer", "https://www.pandalive.co.kr/")
					req.Header.Add("Sec-Fetch-Dest", "empty")
					req.Header.Add("Sec-Fetch-Mode", "cors")
					req.Header.Add("Sec-Fetch-Site", "same-site")
					req.Header.Add("Sec-GPC", "1")
					req.Header.Add("TE", "trailers")

					if err != nil {
						(*user)[2] = "ERROR"
						(*user)[3] = "RETRYING"
						(*user)[4] = err.Error()
						ConcurrentDownload(user, playlist)
					}

					// since panda doesn't actually return a content-length header for go just use this instead of having to read the segments in ram
					filesize, _ := os.Stat("downloads/Panda/" + userId + "/" + filename)
					bytes = filesize.Size()
					elapsed_time := time.Since(start_time)

					(*user)[2] = tools.FormatBytes(bytes)
					(*user)[3] = tools.FormatTime(elapsed_time)
					(*user)[4] = filename

					_, err = io.Copy(out, resp.Body)

					if err != nil {
						(*user)[2] = "ERROR"
						(*user)[3] = "RETRYING"
						(*user)[4] = err.Error()
						ConcurrentDownload(user, playlist)
					}

					playlistUrls[line] = true
				}
			}
		}

		time.Sleep(3 * time.Second)
	}
}

func Download(userId string, playlist string) {
	tools.Exists("downloads/Panda/" + userId)

	client := &http.Client{}

	cookies := tools.LoadCookies("panda")

	req, _ := http.NewRequest("GET", playlist, nil)

	req.Header.Add("Cookie", "sessKey="+cookies[0])
	req.Header.Add("User-Agent", "Mozilla/5.0 (X11; Linux x86_64; rv:124.0) Gecko/20100101 Firefox/124.0")
	req.Header.Add("Accept", "application/json, text/plain, */*")
	req.Header.Add("Accept-Language", "en-US,en;q=0.5")
	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Add("Origin", "https://www.pandalive.co.kr")
	req.Header.Add("DNT", "1")
	req.Header.Add("Connection", "keep-alive")
	req.Header.Add("Referer", "https://www.pandalive.co.kr/")
	req.Header.Add("Sec-Fetch-Dest", "empty")
	req.Header.Add("Sec-Fetch-Mode", "cors")
	req.Header.Add("Sec-Fetch-Site", "same-site")
	req.Header.Add("Sec-GPC", "1")
	req.Header.Add("TE", "trailers")

	filename := userId + "-" + time.Now().Format("200601021504") + "-panda.ts"

	out, _ := os.Create("downloads/Panda/" + userId + "/" + filename)

	playlistUrls := make(map[string]bool)

	var bytes int64 = 0
	var start_time = time.Now()

	for {
		resp, err := client.Do(req)

		if err != nil {
			fmt.Println(err)
		}

		bodyBytes, _ := io.ReadAll(resp.Body)

		bodyText := string(bodyBytes)

		if !strings.Contains(bodyText, ".ts") {
			Start(userId)
		}

		scanner := bufio.NewScanner(strings.NewReader(bodyText))

		for scanner.Scan() {
			line := scanner.Text()
			if strings.HasPrefix(line, "https://") {
				if !playlistUrls[line] {
					req, _ := http.NewRequest("GET", line, nil)

					req.Header.Add("Cookie", "sessKey="+cookies[0])
					req.Header.Add("User-Agent", "Mozilla/5.0 (X11; Linux x86_64; rv:124.0) Gecko/20100101 Firefox/124.0")
					req.Header.Add("Accept", "application/json, text/plain, */*")
					req.Header.Add("Accept-Language", "en-US,en;q=0.5")
					req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
					req.Header.Add("Origin", "https://www.pandalive.co.kr")
					req.Header.Add("DNT", "1")
					req.Header.Add("Connection", "keep-alive")
					req.Header.Add("Referer", "https://www.pandalive.co.kr/")
					req.Header.Add("Sec-Fetch-Dest", "empty")
					req.Header.Add("Sec-Fetch-Mode", "cors")
					req.Header.Add("Sec-Fetch-Site", "same-site")
					req.Header.Add("Sec-GPC", "1")
					req.Header.Add("TE", "trailers")

					resp, err := client.Do(req)

					if err != nil {
						fmt.Println(err)
					}

					// since panda doesn't actually return a content-length header for go just use this instead of having to read the segments in ram
					filesize, _ := os.Stat("downloads/Panda/" + userId + "/" + filename)
					bytes = filesize.Size()
					elapsed_time := time.Since(start_time)

					fmt.Printf("\rDownloading to %s || %s @ %s      \x1b[?25l", filename, tools.FormatTime(elapsed_time), tools.FormatBytes(bytes))

					_, err = io.Copy(out, resp.Body)

					if err != nil {
						fmt.Println(err)
					}

					playlistUrls[line] = true
				}
			}
		}

		time.Sleep(3 * time.Second)
	}
}
